xivo_bus/mixins.py

```diff
@@ -0,0 +1,421 @@
+# Copyright 2021 The Wazo Authors  (see the AUTHORS file)
+# SPDX-License-Identifier: GPL-3.0-or-later
+
+import os
+import sys
+
+from threading import Thread, Lock, Semaphore
+from inspect import isclass
+from six import raise_from, iteritems
+from contextlib import contextmanager
+from collections import namedtuple, defaultdict
+from kombu import Queue, Exchange, Connection, Producer, binding as Binding
+from kombu.mixins import ConsumerMixin as KombuConsumer
+
+from .publishing_queue import PublishingQueue
+from .middlewares import Middleware, MiddlewareError
+
+
+class InitMixin(object):
+    def __init__(
+        self,
+        username='guest',
+        password='guest',
+        host='localhost',
+        port=5672,
+        exchange_name=None,
+        exchange_type=None,
+        middlewares=None,
+        **kwargs
+    ):
+        super(InitMixin, self).__init__(
+            username=username,
+            password=password,
+            host=host,
+            port=port,
+            exchange_name=exchange_name,
+            exchange_type=exchange_type,
+            middlewares=middlewares,
+            **kwargs
+        )
+
+
+class ThreadableMixin(object):
+    def __init__(self, **kwargs):
+        super(ThreadableMixin, self).__init__(**kwargs)
+        self.should_stop = False
+
+    @property
+    def is_running(self):
+        try:
+            return any({thread.is_alive() for thread, _, _ in self.__threads})
+        except AttributeError:
+            return False
+
+    @property
+    def __threads(self):
+        try:
+            return self.__t
+        except AttributeError:
+            self.__t = list()
+            return self.__t
+
+    @property
+    @contextmanager
+    def threads(self):
+        if not self.is_running:
+            self.start()
+        try:
+            yield
+        finally:
+            self.stop()
+
+    def start(self):
+        if self.is_running:
+            raise RuntimeError(
+                '{} threads are already running'.format(len(self.__threads))
+            )
+
+        for thread, _, _ in self.__threads:
+            thread.start()
+
+        for _, _, barrier in self.__threads:
+            barrier.acquire(blocking=True)
+        self.log.debug('All thread started succesfully...')
+
+    def stop(self):
+        self.should_stop = True
+        for thread, cleanup, _ in self.__threads:
+            if thread.is_alive():
+                self.log.info('Stopping AMQP thread \'%s\'', thread.name)
+                if callable(cleanup):
+                    cleanup()
+                thread.join()
+        self.should_stop = False
+
+    def _register_thread(self, name, run, cleanup=None, **kwargs):
+        if not callable(run):
+            raise ValueError('Run must be a function')
+        name = '{}:{}'.format(self._name, name)
+        func = self.__wrap(run)
+        barrier = Semaphore(0)
```

Comment:
Why using Semaphore vs Lock in this context?
It should be acquire only once, non? (if there is a reason, you can call me to *speedup* the explanation :sweat_smile: )

Reply:
Locks need to be acquired and released by the same thread, where semaphore can be post'ed or awaited by any thread.  Semaphore are used in synchronization where locks manages access to resources.

In this case, they replace the barrier from threading library since it's not available in python2.7

Reply:
Not according the documentation: https://docs.python.org/3/library/threading.html#threading.Lock.release

> Release a lock. This can be called from any thread, not only the thread which has acquired the lock.